### ISSUE TITLE
Little explanation + bugfix

### ISSUE THREAD

**Issue by eL-Prova** [defuse/php-encryption#358]
Hi all,

In the source you are using final class Crypto but also a Crypto.
The Crypto (not final) is just a helper, am I correct?
Should it not be better to call it something like CryptoHelper?

Further in Core.class.php. function secureRandom has `self::ensureFunctionExists('random_bytes')` outside the try catch, this should be inside so the exception is correctly handled

**Comment by defuse** [comment]
Thanks for this!

I think you might be looking at an older version of the code. Which version are you looking at? In the latest version I can only find one `Crypto` class (`Crypto.php`) and only one `createNewRandomKey` (inside `Key`). The `KEY_BYTE_SIZE` constant in `Crypto` is also no longer there.

The library uses ParagonIE's `random_compat` library which polyfills `random_bytes` if it isn't available from PHP. It tries hard to find a secure source of randomness, but if none can be found, then it'll have to throw an error.

**Comment by eL-Prova** [comment]
@defuse you are right man...
I will try to find out which version I took..I don't know.
I think I need to review again :p I will throw back my code and will include it again.
If I can help you let me know, anyway I am very happy you take a good and stable repo! Keep on going the hard work :-)

**Comment by defuse** [comment]
Thanks! I appreciate the time you took to look through the code!

**Comment by eL-Prova** [comment]
I used the latest version. Reviewing will I do.
Maybe good to add the previous random as fallback.
So when random_bytes is not available, fall back on the openssl_random_speudo_bytes.

**Comment by paragonie-scott** [comment]
>So when random_bytes is not available, fall back on the openssl_random_speudo_bytes.

That's the thing: Because of [random_compat](https://github.com/paragonie/random_compat), `random_bytes()` is **always** available.

And you don't want to fall back to `openssl_random_pseudo_bytes()`, ever: https://github.com/ramsey/uuid/issues/80

**Comment by defuse** [comment]
Yes, KEY_CURRENT_VERSION is independent from CURRENT_VERSION in Core.php (to differentiate a ciphertext from an encrypted key).

**Comment by cyorobert** [comment]
@defuse I'm not certain I understand what CURRENT_VERSION is doing. Is that constant subject to change in future releases? If so, wouldn't that break the ability to decrypt information from previous versions of the library given the following lines of code?

```php
        // Get and check the version header.
        $header = Core::ourSubstr($ciphertext, 0, Core::HEADER_VERSION_SIZE);
        if ($header !== Core::CURRENT_VERSION) {
            throw new Ex\WrongKeyOrModifiedCiphertextException(
                'Bad version header.'
            );
        }
```

**Comment by defuse** [comment]
Future versions would not only change `CURRENT_VERSION`, they would *also* change those lines of code to detect and support old versions.

**Comment by cyorobert** [comment]
@defuse Interesting, given that this seems to be quite a bit of maintenance overhead, why use this as part of the cipher in the first place?

**Comment by defuse** [comment]
This might help make things more clear: `CURRENT_VERSION` isn't the version of the library, it's the version of the ciphertext format (which is independent of the version of the library, e.g. we might add more features without changing the ciphertext format). It's not expected to ever change, even as more features are added to the library, unless there is some security vulnerability found in the current format.

**Comment by defuse** [comment]
(And in case we do have to change the ciphertext format, it'll be really helpful to have a version header already there, instead of having to somehow implement "if (there's no version header) { do stuff }")

**Comment by cyorobert** [comment]
@defuse Interesting, thank you very much for clarifying! That makes much more sense.
